libs/completer/completer_test.go

```diff
@@ -0,0 +1,113 @@
+package completer
+
+import (
+	"context"
+	"errors"
+	"io/fs"
+	"testing"
+
+	"github.com/databricks/cli/cmd/root"
+	mockfiler "github.com/databricks/cli/internal/mocks/libs/filer"
+	"github.com/databricks/cli/internal/testutil"
+	"github.com/databricks/databricks-sdk-go/experimental/mocks"
+	"github.com/spf13/cobra"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/mock"
+)
+
+func setup(t *testing.T) context.Context {
+	ctx := context.Background()
+	// Needed to make type context.valueCtx for mockFilerForPath
+	ctx = root.SetWorkspaceClient(ctx, mocks.NewMockWorkspaceClient(t).WorkspaceClient)
+	return ctx
+}
+
+func TestFilerCompleterReturnsNestedDirs(t *testing.T) {
+	ctx := setup(t)
+
+	mockFilerForPath := testutil.GetMockFilerForPath(t, "/", []fs.DirEntry{
+		testutil.NewFakeDirEntry("dir", true),
+	})
+	mockFiler, _, _ := mockFilerForPath(ctx, "dbfs:/")
+
+	completer := New(ctx, mockFiler, true)
+
+	completions, dirPath, directive := completer.CompletePath("/")
+
+	assert.Equal(t, []string{"dir"}, completions)
+	assert.Equal(t, "/", dirPath)
+	assert.Equal(t, cobra.ShellCompDirectiveNoSpace, directive)
+}
+
+func TestFilerCompleterReturnsAdjacentDirs(t *testing.T) {
+	ctx := setup(t)
+
+	mockFiler := mockfiler.NewMockFiler(t)
+
+	// First call to get nested dirs fails so we get the adjacent dirs instead
+	mockFiler.On("ReadDir", mock.AnythingOfType("*context.valueCtx"), "/wrong_path").Return(nil, errors.New("error"))
+	mockFiler.On("ReadDir", mock.AnythingOfType("*context.valueCtx"), "/").Return([]fs.DirEntry{
+		testutil.NewFakeDirEntry("adjacent", true),
+	}, nil)
+
+	completer := New(ctx, mockFiler, true)
+	completions, dirPath, directive := completer.CompletePath("/wrong_path")
+
+	assert.Equal(t, []string{"adjacent"}, completions)
+	assert.Equal(t, "/", dirPath)
+	assert.Equal(t, cobra.ShellCompDirectiveNoSpace, directive)
+
+	mockFiler.AssertExpectations(t)
+}
+
+func TestFilerCompleterReadDirError(t *testing.T) {
+	ctx := setup(t)
+
+	mockFiler := mockfiler.NewMockFiler(t)
+	mockFiler.On("ReadDir", mock.AnythingOfType("*context.valueCtx"), mock.Anything).Return(nil, errors.New("error"))
+
+	completer := New(ctx, mockFiler, true)
+	completions, dirPath, directive := completer.CompletePath("/dir")
+
+	assert.Nil(t, completions)
+	assert.Equal(t, "/", dirPath)
+	assert.Equal(t, cobra.ShellCompDirectiveNoSpace, directive)
+
+	mockFiler.AssertExpectations(t)
+}
+
+func TestFilerCompleterReturnsFileAndDir(t *testing.T) {
+	ctx := setup(t)
+
+	mockFilerForPath := testutil.GetMockFilerForPath(t, "/", []fs.DirEntry{
+		testutil.NewFakeDirEntry("dir", true),
+		testutil.NewFakeDirEntry("file", false),
+	})
+	mockFiler, _, _ := mockFilerForPath(ctx, "dbfs:/")
+
+	// onlyDirs=false so we should get both files and directories
+	completer := New(ctx, mockFiler, false)
+
+	completions, dirPath, directive := completer.CompletePath("/")
```

Comment:
Does autocompletion work if user types in `f`? It seems like completions will not return `file` in this case?Similar to the sync autocomplete we should consider using `path.Dir` to provide path completion.

Reply:
Yup it works:
<img width="566" alt="Screenshot 2024-08-06 at 8 39 51" src="https://github.com/user-attachments/assets/a742b122-e447-427a-a0e1-b2e87c5ae951">

Reply:
Yup it works